structuredheadings/plugin.js

```diff
@@ -222,24 +183,162 @@
+
+        onClick: function (value) {
+          editor.applyStyle(elementStyles[ value ]);
+          setLevel(editor, editor.elementPath().block);
+          editor.execCommand("reapplyStyle");
+          if (value === "p") {
+            this.setValue(value, "Normal Text");
+          } else if (value === "pre") {
+            this.setValue(value, "Formatted Text");
           } else {
-            this.setState(CKEDITOR.TRISTATE_OFF);
+            this.setValue(value, "Header " + editor.config.numberedElements[
+                editor.config.numberedElements.indexOf(value)
+              ].slice(1));
           }
+
+        },
+
+        onRender: function () {
+          editor.on("selectionChange", function (ev) {
+            var currentTag = this.getValue();
+            var elementPath = ev.data.path;
+            var elementList = editor.config.numberedElements.concat(["pre", "p", "div"]);
+            for (var tag in elementList) {
+              if (elementStyles[elementList[tag]].checkActive(elementPath, editor)) {
+                if (elementList[tag] !== currentTag) {
+                  if (elementList[tag] === "p") {
+                    this.setValue(elementList[tag], "Normal Text");
+                  } else if (elementList[tag] === "pre") {
+                    this.setValue(elementList[tag], "Formatted Text");
+                  } else {
+                    this.setValue(elementList[tag], "Header " +
+                          editor.config.numberedElements[tag].slice(1));
+                  }
+                }
+                return;
+              }
+            }
+
+            this.setValue("");
+          }, this);
         }
-      }));
 
-      CKEDITOR.dialog.add("selectStyle", this.path + "dialogs/selectstyle.js");
+      });
+
+      //Style Dropdown
+      editor.ui.addRichCombo("NumStyles", {
+        label: "Numbering",
+        title: "Numbering Styles",
+        toolbar: "styles,8",
+        allowedContent: "h1(*); h2(*); h3(*); h4(*); h5(*); h6(*)",
+
+        panel: {
+          css: [ CKEDITOR.skin.getPath("editor") ].concat(editor.config.contentsCss),
+          multiSelect: false,
+          attributes: { "aria-label": "Numbering Styles" }
+        },
+
+        init: function () {
+          this.startGroup("Styles");
+
+          for (var style in editor.config.autonumberStyles) {
+            this.add(style, style, style);
+          }
+
+          this.startGroup("Setup");
+          this.add("clear", "Clear Numbering", "Clear Numbering");
+          this.add("restart", "Restart Numbering", "Restart Numbering");
+        },
+
+        onClick: function (value) {
+          if (value === "restart") {
+            editor.execCommand("restartNumbering");
+          } else if (value === "clear") {
+            clearStyles(editor, editor.elementPath().block);
+            clearLevel(editor, editor.elementPath().block);
+            clearNumbering(editor, editor.elementPath().block);
+          } else {
```

Comment:
restart and clear are hardcoded entries to the select element with specific commands, the else accounts for each of the stylenames which are dynamic, based on config setting and applies that style.